### DISCUSSION TITLE
Player teleport no chunk

### DISCUSSION THREAD

**Discussion by matserix** [Minestom/Minestom#1529] [Help]
When I teleport the player to a chunk which he doesn't see, the chunk won't get shown to the player. The chunks are loaded on server-side. Is there a way to fix this? Thanks in advance

I'm using the anvil loader btw 

![2022-11-20_16 59 17](https://user-images.githubusercontent.com/113344017/202912415-2ac777d1-0bd3-4fdc-8946-6bbe6302dacf.png)

**Comment by KrystilizeNevaDies** [comment]
How are you teleporting the player? Do you do anything before the player has finished being teleported?
Can you also confirm that the chunk packet is being sent?

**Comment by matserix** [comment]
This is my code for teleport and the packet does not get send

```kotlin
when (slot) {
    20 -> player.teleport(Locations.CITY_BUILD)
    22 -> player.teleport(Locations.SPAWN)
    else -> return
}
player.closeInventory()
```

**Reply by KrystilizeNevaDies** [reply]
Player#teleport returns a completable future, could you try joining those futures?

**Reply by Obyvante** [reply]
```java
                final var chunk = player.getInstance().getChunkAt(found_position);
                chunk.sendChunk(player);
                player.teleport(found_position).thenAccept(v -> chunk.sendChunk(player));
```
This isn't also working. I tried to send the packet before and after. I will try to send packet and then wait for next tick to teleport next.
_Ignore the code style and logic._

**Reply by matserix** [reply]
If you mean that I should use .join() then it does nothing, no exception. I also found out that the first teleport does function correctly when the chunk wasn't already sent to the player (Player didn't see it yet)

**Reply by KrystilizeNevaDies** [reply]
It seems like the player is never fully teleported, and hence the future is never completed.

Try stepping through the teleport method in debug with breakpoints to see what is stopping the teleportation from happening.

**Reply by matserix** [reply]
After trying out the try of @Obyvante I didn't fell into void anymore. I could walk in the air and when I left the chunk I was teleported to everything was visible

**Reply by Obyvante** [reply]
> After trying out the try of @Obyvante I didn't fell into void anymore. I could walk in the air and when I left the chunk I was teleported to everything was visible

![image](https://user-images.githubusercontent.com/45916622/203614875-b793eef2-7bad-49a9-bf64-654af8444b40.png)
Yeah, it seems like we have to send a chunk packet again or send a chunk packet before sending a player to the chunk. 

Seems it works with adding some delay. @NitrinCloud Could you please try to reproduce the issue with the below code?
```java
                final var player = (Player) sender;
                final var found_position = Dimension.INSTANCE.findValidPosition();
                
                player.getInstance().getChunkAt(found_position).sendChunk(player);
                MinecraftServer.getSchedulerManager().buildTask(() -> player.teleport(found_position))
                        .delay(TaskSchedule.nextTick())
                        .schedule();
```

**Reply by matserix** [reply]
> > After trying out the try of @Obyvante I didn't fell into void anymore. I could walk in the air and when I left the chunk I was teleported to everything was visible
> 
> ![image](https://user-images.githubusercontent.com/45916622/203614875-b793eef2-7bad-49a9-bf64-654af8444b40.png) Yeah, it seems like we have to send a chunk packet again or send a chunk packet before sending a player to the chunk.
> 
> Seems it works with adding some delay. @NitrinCloud Could you please try to reproduce the issue with the below code?
> 
> ```java
>                 final var player = (Player) sender;
>                 final var found_position = Dimension.INSTANCE.findValidPosition();
>                 
>                 player.getInstance().getChunkAt(found_position).sendChunk(player);
>                 MinecraftServer.getSchedulerManager().buildTask(() -> player.teleport(found_position))
>                         .delay(TaskSchedule.nextTick())
>                         .schedule();
> ```

```kotlin
override fun handleInput(player: Player, slot: Int, clickType: ClickType, itemStack: ItemStack) {
        val spawnChunk = Lobby.instance.getChunkAt(Locations.CITY_BUILD)
        spawnChunk?.sendChunk(player)

        val cityChunk = Lobby.instance.getChunkAt(Locations.CITY_BUILD)
        cityChunk?.sendChunk(player)
        MinecraftServer.getSchedulerManager().buildTask { when (slot) {
            20 -> player.teleport(Locations.CITY_BUILD)
            22 -> player.teleport(Locations.SPAWN)
            else -> {}
        }}.delay(TaskSchedule.nextTick()).schedule()
        player.closeInventory()
    }
```

Not workin, still the same as with your screenshot

**Reply by Obyvante** [reply]
Could you try to change the tick to 5 or more to see if the client is got the chunk packet?

**Reply by matserix** [reply]
Time doesnt matter, tried a few ticks and seconds

**Reply by matserix** [reply]
> It seems like the player is never fully teleported, and hence the future is never completed.
> 
> Try stepping through the teleport method in debug with breakpoints to see what is stopping the teleportation from happening.

The problem disappears when it is in the chunk-view-distance. When print out isDone from the teleport on the first time it says false, this is the time the teleport works. On every following teleport the isDone is true and it doesnt work

**Reply by matserix** [reply]
> Could you try to change the tick to 5 or more to see if the client is got the chunk packet?

Your code only works, because your position is random and is not already generated. You don't use a position twice. First time teleport always works, every teleport after doesnt

**Comment by matserix** [comment]
```java
public static void main(String[] args) {
        System.setProperty("minestom.chunk-view-distance", "6");

        var server = MinecraftServer.init();

        InstanceContainer instance = MinecraftServer.getInstanceManager().createInstanceContainer();
        instance.setGenerator(unit -> unit.modifier().fillHeight(0, 40, Block.GRASS_BLOCK));

        MinecraftServer.getGlobalEventHandler().addListener(PlayerLoginEvent.class, event -> {
            event.setSpawningInstance(instance);
            event.getPlayer().setRespawnPoint(new Pos(0.0, 42.0, 0.0));
        });

        Command command = new Command("test");

        final boolean[] check = {false};
        command.setDefaultExecutor((sender, context) -> {
            if (sender instanceof Player player) {
                if (check[0]) {
                    player.teleport(new Pos(10000.0, 42.0, 10000.0));
                } else {
                    player.teleport(new Pos(0.0, 42.0, 0));
                }
                check[0] = !check[0];
            }
        });
        MinecraftServer.getCommandManager().register(command);

        server.start(new InetSocketAddress("127.0.0.1", 25565));
    }
```

This problem seems to be easy to reproduce. Would appreciate if anyone found a workarround or fix the "bug" in minestom code itself, I'm not able to do anything myself anymore
